src/planner/binder.cpp

```diff
@@ -512,11 +512,26 @@ void Binder::AddTableName(string table_name) {
 	root_binder.table_names.insert(std::move(table_name));
 }
 
+void Binder::AddReplacementScan(const string &table_name, unique_ptr<TableRef> replacement) {
+	auto &root_binder = GetRootBinder();
+	auto it = root_binder.replacement_scans.find(table_name);
+	if (it == root_binder.replacement_scans.end()) {
+		root_binder.replacement_scans[table_name] = std::move(replacement);
+	} else {
+		// A replacement scan by this name was previously registered, we can just use it
```

Comment:
I'm not entirely sure about this
If we create an object in the replacement scan every time, but we only keep the first one, then the second replacement will be corrupt

We should probably use this map as a cache so we don't run into this scenario - so we don't create the replacement scan a second time

Reply:
Actually that's not what happens

It doesn't corrupt, all instances of `df` are just replaced with the first replacement scan
so what ends up happening is that because things like pyarrow RecordBatchReaders are destructive, everything except for the first scan becomes empty